### PR TITLE
Strtree docs

### PR DESCRIPTION
**Pull request by Jeremiah-England**
I've update the docs for the STRtree in manual.rst. This should close these issues (#854, #845, #618) and one other pull request (https://github.com/Toblerity/Shapely/pull/623). Note that I have not done anything with the docstrings. I think I would like to work on that in a separate merge later. Maybe waiting to close those issues until that is done as well is a good idea. 

Also note that I have changed the work around for getting indexes to use an auxiliary dictionary keyed on geometry memory IDs instead of adding an index attribute to the geometries themselves (or a deep copy if you don't want to mutate.) 

Compare: 

``` python 
>>> from copy import deepcopy
>>> 
>>> points = [Point(i, i) for i in range(10)]
>>> tree = STRtree(points)
>>>
>>> points_numbered = []
>>> for i, point in enumerate(points):
>>>     point = deepcopy(point)  # let's not mutate the originals
>>>     point.i = i
>>>     points_numbered.append(point)
>>> tree = STRtree(points_numbered)
>>> result = tree.query(Point(2, 2).buffer(1.0))
>>> [(o.i, o.wkt) for o in result]
[(1, 'POINT (1 1)'), (2, 'POINT (2 2)'), (3, 'POINT (3 3)')]
```  
vs 

``` python 
>>> points = [Point(i, i) for i in range(10)]
>>> tree = STRtree(points)
>>>
>>> index_by_id = dict((id(pt), i) for i, pt in enumerate(points))
>>> [(index_by_id[id(pt)], pt.wkt) for pt in tree.query(Point(2,2).buffer(1.0))]
[(1, 'POINT (1 1)'), (2, 'POINT (2 2)'), (3, 'POINT (3 3)')]
```

This appears to be robust even in the case where the list grows, which you might expect to change memory location (id) of the geometries.  Observe:

``` python
>>> points = [Point(-1, -1)]
>>> original_id = id(points[0])
>>> for i in range(1000000):
>>>     points.append([Point(i, i)])
>>>     if id(points[0]) != original_id:
>>>         print(f"Element id changed from {original_id} to {id(points[0])}")
>>>         break
>>> original_id == id(points[0])
True
```

The RST rendering has been tested. Results are available here: https://jeremiah-england.github.io/shapely/build/manual.html#str-packed-r-tree  

Any review and feedback will be appreciated.

### CODE REVIEW

File: docs/manual.rst [hunk 1]
```diff
@@ -2300,9 +2300,9 @@ The :func:`~shapely.ops.substring` function in `shapely.ops` returns a line segm
     Negative distance values are taken as measured in the reverse
     direction from the end of the geometry. Out-of-range index
     values are handled by clamping them to the valid range of values.
-    
+
     If the start distance equals the end distance, a point is being returned.
-
```
Comment (sgillies) [inline docs/manual.rst]: 👍

File: docs/manual.rst [hunk 2]
```diff
@@ -2409,37 +2409,62 @@ STR-packed R-tree
 
 Shapely provides an interface to the query-only GEOS R-tree packed using the
 Sort-Tile-Recursive algorithm. Pass a list of geometry objects to the STRtree
-constructor to create an R-tree that you can query with another geometric object.
+constructor to create a spatial index that you can query with another geometric
+object. Query-only means that once created, the `STRtree` is immutable. You
+cannot add or remove geometries.
```
Comment (sgillies) [inline docs/manual.rst]: 👍

File: docs/manual.rst [hunk 3]
```diff
@@ -2409,37 +2409,62 @@ STR-packed R-tree
...
 
 .. class:: strtree.STRtree(geometries)
 
   The `STRtree` constructor takes a sequence of geometric objects.
 
-  These are copied and stored in the R-tree.
+  These are copied and stored in the R-tree. However, references are kept to
+  the original geometries and queries return these, not copies.
```
Comment (sgillies) [inline docs/manual.rst]: Let's delete the original text, which is in fact erroneous, and say something like "references to these geometric objects are stored in the R-tree".

File: docs/manual.rst [hunk 4]
```diff
@@ -2409,37 +2409,62 @@ STR-packed R-tree
...
-intersect the provided geometry argument. If you want to match geometries of a
-more specific spatial relationship (eg. crosses, contains, overlaps), consider
-performing the query on the R-tree, followed by a manual search through the
-returned subset using the desired binary predicate.
+  .. method:: strtree.query(goem)
 
-Query-only means that once created, the R-tree is immutable. You cannot 
-add or remove geometries.
+    Returns a list of all geometries in the `strtree` whose extents intersect the
+    extents of `goem`. This means that a subsequent search through the returned
```
Comment (sgillies) [inline docs/manual.rst]:
```suggestion
    extents of `geom`. This means that a subsequent search through the returned
```

File: docs/manual.rst [hunk 5]
```diff
@@ -2409,37 +2409,62 @@ STR-packed R-tree
...
+    subset using the desired binary predicate (eg. intersects, crosses, contains,
+    overlaps) may be necessary to further filter the results according to their
+    specific spatial relationships.
```
Comment (sgillies) [inline docs/manual.rst]: 👍

File: docs/manual.rst [hunk 6]
```diff
@@ -2409,37 +2409,62 @@ STR-packed R-tree
...
+      .. code-block:: pycon
+
+        >>> index_by_id = dict((id(pt), i) for i, pt in enumerate(points))
+        >>> [(index_by_id[id(pt)], pt.wkt) for pt in tree.query(Point(2,2).buffer(1.0))]
+        [(1, 'POINT (1 1)'), (2, 'POINT (2 2)'), (3, 'POINT (3 3)')]
+
+
+  .. method:: strtree.nearest(geom)
+
+    Returns the nearest goemetry in `strtree` to `geom`.
```
Comment (sgillies) [inline docs/manual.rst]:
```suggestion
    Returns the nearest geometry in `strtree` to `geom`.
```

File: docs/manual.rst (hunk 3 above)
Comment (Jeremiah-England) [reply docs/manual.rst]: af89f54000e95ad99095e64b92082fc751fd4aab

**Comment by Jeremiah-England** [comment]
@sgillies I've fixed the problems you mentioned. Let me know if you find any others. Thanks.

**Comment by kannes** [comment]
Great work!

**Pull request merged**